### ISSUE TITLE
Concurrent::Promises.future deadlock in RSpec test

### ISSUE THREAD

**Issue by chernenkov-ayu-sbmt** [ruby-concurrency/concurrent-ruby#1055]
Hello! I found deadlock behaviour when I was writing RSpec test for my custom `Scientist::Experiment` implementation (see [scientist](https://github.com/github/scientist) gem) which use `Concurrent::Promises.future` for running observations in parallel. I'm not sure why it happens and how to resolve it. I prepared a [small example code](https://github.com/chernenkov-ayu-sbmt/concurrent-ruby-rspec-hangs) which reproduces the deadlock. Please, help!

It is interesting, that when I change `let(:experiment_run)` to:

```ruby
Scientist.run "test_experiment" do |e|
  e.use { "expected_result" }
  e.try { "expected_result" }
end
```

then test finishes successfully.

It is also interesting, that in my original code (for the Rails service) doesn't raise `Deadlock` error but hangs forever instead.

Context:

```
* Operating system:                mac
* Ruby implementation:             Ruby 3.3.0
* `concurrent-ruby` version:       1.3.3
* `concurrent-ruby-ext` installed: no
* `concurrent-ruby-edge` used:     no
```

**Comment by bensheldon** [comment]
Thanks for opening the issue, and very much for the nice reproduction. I don't think the source of this problem is Concurrent Ruby, nor Scientist

I checked out your example. I think the cause is from RSpec `let` blocks that [have a Mutex](https://github.com/rspec/rspec-core/blob/4be46f6e7839692f7d1ffd8d585c46a0f5f1e15e/lib/rspec/core/memoized_helpers.rb#L173) and are lazy evaluated. The simplest change to your code would be to use `let!` so they are eager-evaluated.

I don't know enough about RSpec to say whether deadlock on `let` is expected, undefined, or a bug, but you might search RSpec issues to see if this has been reported and if not, do so. I found it interesting.
